### PR TITLE
Monolithic tracer systems (take 2)

### PR DESCRIPTION
**Pull request by jwallwork23**
I finally got round to reimplementing #268 following the recommended changes.

The previous version introduced unnecessary special time integrator classes. In this implementation, the 2D tracer equation formulations are unified and multiple tracer equations can be solved together as a monolithic system. I added a new demo `gray_scott_mixed.py` to check that it works and the results are visually identical to those of `gray_scott.py`, which solves two tracer systems alternately using two Picard iterations per timestep.

Might be good to discuss in Friday's Thetis meeting.

### CODE REVIEW

**Review by mc4117 — COMMENTED** [review]
This looks good but I am just a bit confused why the test/sediment files disappearing as part of this pull request?

**Comment by jwallwork23** [comment]
I don't think they disappear. I just noticed that they were executable for some reason, so applied `chmod -x`.

**Comment by mc4117** [comment]
That makes sense - thanks!

**Review by mc4117 — DISMISSED** [review]

File: thetis/tracer_eq_2d.py [hunk 1]
```diff
@@ -261,43 +292,210 @@ class SourceTerm(TracerTerm):
...
+            H = self.depth.get_total_depth(fields_old['elev_2d'])
+            f += -inner(H*source, self.test)*self.dx
+        return -f
+
+
 class TracerEquation2D(Equation):
     """
-    2D tracer advection-diffusion equation :eq:`tracer_eq` in conservative form
+    Several 2D tracer advection-diffusion equations :eq:`tracer_eq` in
+    conservative or non-conservative form, solved as a monolithic system.
```
Comment (tkarna) [inline thetis/tracer_eq_2d.py:451]: mixed system?

**Review by tkarna — DISMISSED** [review]
Looks good, nice refractoring in the tracer equation. Essentially my only comment is to find another name for `add_tracers_2d`.

File: thetis/options.py [hunk 2]
```diff
@@ -912,6 +914,38 @@ def add_tracer_2d(self, label, name, filename, shortname=None, unit='-', **kwarg
         self.tracer[label].source = kwargs.get('source')
         self.tracer[label].diffusivity = kwargs.get('diffusivity')
         self.tracer[label].use_conservative_form = kwargs.get('use_conservative_form', False)
+        if not kwargs.get('mixed', False):
+            self.tracer_fields[label] = self.tracer[label].function
+
+    def add_tracers_2d(self, labels, names, filenames, shortnames=None, units=None, function=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply thetis/options.py:920]: Ah yes, good suggestion

**Review by tkarna — DISMISSED** [review]
Thanks for the updates!

**Comment by jwallwork23** [comment]
Thanks! Just rebasing on top of master

**Review by tkarna — APPROVED** [review]

**Pull request merged**